### PR TITLE
macOS: request camera & microphone permissions on demand

### LINKED ISSUE #109: Camera and microphone not working on macOS
### Expected Behavior

Attempting to use the video sensing extension or sound recorder should, on macOS 10.14 and higher:
- Request permission to use the camera/microphone (at least the first time)
- Cause the application to appear in the Camera or Microphone section of "Security & Privacy" in System Preferences
- Function correctly in the application

### Actual Behavior

On macOS 10.14.6, none of the above items are true.

### Steps to Reproduce

Camera:
1. Run Scratch Desktop 3.9.0 or higher
2. Add the "Video Sensing" extension
3. Look at the stage

Microphone:
1. Run Scratch Desktop 3.9.0 or higher
2. Go to the "Sounds" tab and activate the sound recorder
3. Note that the loudness indicator does not correctly respond to noise
4. Press "Record"
5. Note that only silence is recorded

### Operating System and Browser

macOS 10.14.6
Scratch Desktop 3.9.0 (Electron 6.1.7, Chrome 76.0.3809.146)

### PR DESCRIPTION
**Pull request by cwillisf**
### Resolves

- Resolves #109 

### Proposed Changes

This batch of changes uses `askForMediaAccess` to request camera and microphone access when necessary.

Supporting changes include:

- Add a section in `README.md` mentioning a known code-signing issue which can sometimes interfere with debugging
- Use separate entitlements lists for Mac App Store vs. non-store ("Developer ID") macOS builds
  - Both macOS builds now use the `cs.allow-jit` entitlement since documentation suggests that it's necessary for JavaScript to work well.
  - non-Store mac builds no longer use the `app-sandbox` entitlement since it seems to cause code-signing problems. Mac App Store builds still use it as it's required for the store.
- Add descriptions for why the app asks for camera and microphone permissions, the consequences of denying those permissions, and how to change your mind.

### Reason for Changes

On macOS 10.14 (Mojave) and macOS 10.15 (Catalina), apps must request access to the microphone or camera. Previous versions of Scratch Desktop didn't request permission, so camera and microphone features didn't work correctly.

### Test Coverage

Manually tested. Example test sequences below.

Note that the `tccutil reset Camera` and/or "tccutil reset Microphone` command(s) must be used before each test if you would like to see the prompt again. You can also enable or disable permission through System Preferences.

Camera:
- Run `tccutil reset Camera` (this will reset camera permissions for ALL applications)
- Start this build of Scratch Desktop
- Attempt to add a costume by taking a picture, or add the "video sensing" extension
- You should get a pop-up dialog asking for permission
   - If you deny the access the "video sensing" extension and sound recorder won't work. You should receive a dialog explaining why the application needs these features.
   - If you allow the access the video sensing extension and taking a photo in the paint editor should work correctly.

Microphone:
- Run `tccutil reset Microphone` (this will reset microphone permissions for ALL applications)
- Start this build of Scratch Desktop
- Attempt to record a sound in the sound editor or add the "loudness" block to the stage
- You should get a pop-up dialog asking for permission
   - If you deny the access the loudness block and sound recorder won't work. You should receive a dialog explaining why the application needs these features.
   - If you allow the access the loudness block and sound recorder should work correctly.

### CODE REVIEW

File: src/main/index.js [hunk 1]
```diff
@@ -18,22 +18,48 @@ const isDevelopment = process.env.NODE_ENV !== 'production';
...
-    });
-    const webContents = window.webContents;
+const displayPermissionDeniedWarning = (browserWindow, permissionType) => {
+    let title;
+    let message;
+    switch (permissionType) {
+    case 'camera':
+        title = 'Camera Permission Denied';
+        message = 'Permission to use the camera has been denied. ' +
+            'Scratch will not be able to take a photo or use video sensing blocks.';
```
Comment (adroitwhiz) [inline src/main/index.js:28]: Will these be localized?

Comment (apple502j) [reply src/main/index.js:28]: Currently, NO strings are translated in scratch-desktop. Maybe in next PR?

Comment (cwillisf) [reply src/main/index.js:28]: Not yet, but localization is definitely on my list for both scratch-desktop and scratch-link. In terms of "now" or "next" or "later" I would say localization in scratch-desktop is currently filed under "next" :)

**Review by paulkaplan — DISMISSED** [review]
Changes LGTM. Glad that setPermissionRequestHandler seems straightforward to use. Curious about the look of that on Windows?

**Comment by cwillisf** [comment]
Oops -- I accidentally removed the `if (process.platform === 'darwin')` part of this! Electron's `systemPreferences` API is only available on macOS, so this definitely needs to be inside a platform check. Changes coming soon...

**Comment by cwillisf** [comment]
Instead of checking the `process.platform` I decided to check for the existence of `systemPreferences.askForMediaAccess` since that feels a bit more future-proof. Windows actually has a similar permissions model now so it wouldn't surprise me if Electron adds support for it someday.

File: src/main/index.js [hunk 2]
```diff
@@ -75,6 +75,23 @@ const makeFullUrl = (url, search = null) =>
...
+ * Any application-level checks, such as whether or not a particular frame or document should be allowed to ask,
+ * should be done before calling this function.
+ *
+ * @param {string} mediaType - one of Electron's media types, like 'microphone' or 'camera'
+ * @returns {boolean} - true if permission granted, false otherwise.
+ */
+const askForMediaAccess = async mediaType => {
+    if (systemPreferences.askForMediaAccess) {
+        // Electron currently only implements this on macOS
+        return await systemPreferences.askForMediaAccess(mediaType);
```
Comment (adroitwhiz) [inline src/main/index.js]: nit: `return await` [is unnecessary](https://eslint.org/docs/rules/no-return-await)
```suggestion
        return systemPreferences.askForMediaAccess(mediaType);
```

Comment (cwillisf) [reply src/main/index.js]: Yeah, I felt a bit weird writing that but I figured it was convention. Looks like I was wrong -- thanks for letting me know about that eslint rule! We should probably add it to `eslint-config-scratch`...

**Review by paulkaplan — APPROVED** [review]
LGTM

**Pull request merged**